src/libraries/System.Private.CoreLib/src/System/Reflection/Emit/TypeNameBuilder.cs

```diff
@@ -0,0 +1,429 @@
+            }
+        }
+
+        private void EscapeName(string name)
+        {
+            if (ContainsReservedChar(name))
+            {
+                foreach (char c in name)
+                {
+                    if (c == '\0')
+                        break;
+                    if (IsTypeNameReservedChar(c))
+                        _str.Append('\\');
+                    _str.Append(c);
+                }
+            }
+            else
+                Append(name);
+        }
+
+        private void EscapeAssemblyName(string name)
+        {
+            Append(name);
+        }
+
+        private void EscapeEmbeddedAssemblyName(string name)
+        {
+            bool containsReservedChar = false;
+
+            foreach (char c in name)
+            {
+                if (c == ']')
+                {
+                    containsReservedChar = true;
+                    break;
+                }
+            }
+
+            if (containsReservedChar)
+            {
+                foreach (char c in name)
+                {
+                    if (c == ']')
+                        Append('\\');
+
+                    Append(c);
+                }
+            }
+            else
+            {
+                Append(name);
+            }
+        }
+
+        private void CheckParseState(ParseState validState)
+        {
+            Debug.Assert((_parseState & validState) != 0);
+        }
+
+        private void PushOpenGenericArgument()
+        {
+            _stack.Add(_str.Length);
+            _stackIdx++;
+        }
+
+        private void PopOpenGenericArgument()
+        {
+            int index = _stack[--_stackIdx];
+            _stack.RemoveAt(_stackIdx);
+
+            if (!_hasAssemblySpec)
+                _str.Remove(index - 1, 1);
+
+            _hasAssemblySpec = false;
+        }
+
+        private void SetUseAngleBracketsForGenerics(bool value)
+        {
+            _useAngleBracketsForGenerics = value;
+        }
+
+        private void Append(string pStr)
+        {
+            foreach (char c in pStr)
+            {
+                if (c == '\0')
+                    break;
+                _str.Append(c);
+            }
+        }
+
+        private void Append(char c)
+        {
+            _str.Append(c);
+        }
+
+        internal enum Format
+        {
+            ToString,
+            FullName,
+            AssemblyQualifiedName,
+        }
+
+        internal static string? ToString(Type type, Format format)
+        {
+            if (format == Format.FullName || format == Format.AssemblyQualifiedName)
+            {
+                if (!type.IsGenericTypeDefinition && type.ContainsGenericParameters)
+                    return null;
+            }
+
+            var tnb = new TypeNameBuilder();
+            tnb.AddAssemblyQualifiedName(type, format);
+            return tnb.ToString();
+        }
+
+        private void AddElementType(Type elementType)
```

Comment:
```suggestion
        private void AddElementType(Type type)
```

This argument is not the element type. It is the type which element type that this is adding.